### PR TITLE
interfaces,interfaces/udev: add udev security backend

### PR DESCRIPTION
**Pull request by zyga**
This branch adds the last security backend, udev, and eradicates all traces of securityHelper code and all the corresponding older versions of the implementation.

### CODE REVIEW

File: interfaces/udev/backend.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+// Deconfigure removes udev rules specific to a given snap.
+// If any of the rules are removed then udev database is reloaded.
+//
+// This method should be called after removing a snap.
+func (b *Backend) Deconfigure(snapInfo *snap.Info) error {
+	glob := fmt.Sprintf("70-%s.rules", interfaces.SecurityTagGlob(snapInfo))
+	changed, removed, err := osutil.EnsureDirState(dirs.SnapUdevRulesDir, glob, nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize udev rules for snap %q: %s", snapInfo.Name, err)
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: Per prior note, what's the value of the partial result if we're ignoring it? Also, what's the plan for fixing the bad data?

File: interfaces/udev/backend.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		return ReloadRules()
+	}
+	return nil
+}
+
+// combineSnippets combines security snippets collected from all the interfaces
+// affecting a given snap into a content map applicable to EnsureDirState.
+func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+	for _, appInfo := range snapInfo.Apps {
+		if len(snippets[appInfo.Name]) > 0 {
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: Same note as for the dbus backend.

File: interfaces/udev/backend.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+			var buf bytes.Buffer
+			buf.WriteString("# This file was automatically generated by snappy\n")
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: Per prior review point, `"# This file is automatically generated.\n"`

(it doesn't help to say it's generated by the operating system)

**Comment by niemeyer** [comment]
A few simple details. Would like to have a look after these are in.

File: interfaces/udev/backend.go (hunk 2 above)
Comment (zyga) [inline interfaces/udev/backend.go]: Ack, I'll write the required tests

File: interfaces/udev/backend.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	var errReload error
+	changed, removed, errEnsure := osutil.EnsureDirState(dir, glob, content)
+	if len(changed) > 0 || len(removed) > 0 {
+		// Try reload the rules regardless of errEnsure.
+		errReload = ReloadRules()
+	}
+	if errEnsure != nil {
+		return fmt.Errorf("cannot synchronize udev rules for snap %q: %s", snapInfo.Name, errEnsure)
+	}
+	if errReload != nil {
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: This should be simply `return errReload`. The line below is repeating the message of ReloadRules itself.

File: interfaces/udev/backend.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+// combineSnippets combines security snippets collected from all the interfaces
+// affecting a given snap into a content map applicable to EnsureDirState.
+func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+	for _, appInfo := range snapInfo.Apps {
+		appSnippets := snippets[appInfo.Name]
+		if len(appSnippets) == 0 {
+			continue
+		}
+		var buf bytes.Buffer
+		buf.WriteString("# This file was automatically generated by snappy\n")
```
Comment (niemeyer) [inline interfaces/udev/backend.go]: Review comment still unhandled.

File: interfaces/udev/backend.go (hunk 4 above)
Comment (zyga) [inline interfaces/udev/backend.go]: Good catch, fixed.

File: interfaces/udev/backend.go (hunk 3 above)
Comment (zyga) [inline interfaces/udev/backend.go]: Fixed

File: interfaces/udev/backend.go (hunk 5 above)
Comment (zyga) [inline interfaces/udev/backend.go]: Fixed now, sorry for missing it.

**Comment by niemeyer** [comment]
I'm merging this. It has the same structure of every other backend and it's the simplest of them.

**Pull request merged**